crates/sui-move-build/src/linters/self_transfer.rs

```diff
@@ -0,0 +1,213 @@
+        diags
+    }
+
+    fn start_command(&self, _: &mut State) -> ExecutionContext {
+        ExecutionContext {
+            diags: Diagnostics::new(),
+        }
+    }
+
+    fn finish_command(&self, context: ExecutionContext, _state: &mut State) -> Diagnostics {
+        let ExecutionContext { diags } = context;
+        diags
+    }
+
+    fn exp_custom(
+        &self,
+        _context: &mut ExecutionContext,
+        _state: &mut State,
+        _e: &Exp,
+    ) -> Option<Vec<Value>> {
+        None
+    }
+
+    fn call_custom(
+        &self,
+        context: &mut ExecutionContext,
+        _state: &mut State,
+        loc: &Loc,
+        return_ty: &Type,
+        f: &ModuleCall,
+        args: Vec<Value>,
+    ) -> Option<Vec<Value>> {
+        if TRANSFER_FUNCTIONS
+            .iter()
+            .any(|(addr, module, fun)| f.is(addr, module, fun))
+            && args[1] == Value::SenderAddress
+        {
+            let msg = format!(
+                "Transfer of an object to `tx_context::sender()` in function {}",
+                self.fn_name
+            );
+            let uid_msg = "Returning an object from a function, allows a caller to use the object and enables composability via programmable transactions.";
+            let d = diag!(SELF_TRANSFER_DIAG, (*loc, msg), (self.fn_ret_loc, uid_msg));
+            context.add_diag(d);
+            return Some(vec![]);
+        }
+        if f.is("sui", "tx_context", "sender") {
+            return Some(vec![Value::SenderAddress]);
+        }
+        Some(match &return_ty.value {
+            Type_::Unit => vec![],
+            Type_::Single(_) => vec![Value::Other],
+            Type_::Multiple(types) => vec![Value::Other; types.len()],
+        })
+    }
+
+    fn command_custom(&self, _: &mut ExecutionContext, _: &mut State, _: &Command) -> bool {
+        false
+    }
+
+    fn lvalue_custom(
+        &self,
+        _context: &mut ExecutionContext,
+        _state: &mut State,
+        _l: &LValue,
+        _value: &Value,
+    ) -> bool {
+        false
+    }
+}
+
+impl SimpleDomain for State {
+    type Value = Value;
+
+    fn new(context: &CFGContext, mut locals: BTreeMap<Var, LocalState<Value>>) -> Self {
+        for (v, _) in &context.signature.parameters {
+            let local_state = locals.get_mut(v).unwrap();
+            if let LocalState::Available(loc, _) = local_state {
+                *local_state = LocalState::Available(*loc, Value::Other);
+            }
+        }
+        State { locals }
+    }
+
+    fn locals_mut(&mut self) -> &mut BTreeMap<Var, LocalState<Value>> {
+        &mut self.locals
+    }
+
+    fn locals(&self) -> &BTreeMap<Var, LocalState<Value>> {
+        &self.locals
+    }
+
+    fn join_value(v1: &Value, v2: &Value) -> Value {
+        match (v1, v2) {
+            (Value::SenderAddress, Value::SenderAddress) => Value::SenderAddress,
+            (Value::Other, _) | (_, Value::Other) => Value::Other,
```

Comment:
How noisy do we want this to be? 
If we did `(Other, SenderAddress) => SenderAddress`, we could catch more cases. But I could definitely see some of those cases being legitimate. On the flip side, they might want to rethink their code overall (maybe). Thoughts?

Reply:
I'd rather go for less noisy version - if we find out that it's not catching enough cases, we can always revisit.